src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+#define BUFFER_LENGHT_DEVICE_PATH "/sys/bus/iio/devices/iio:device%d/buffer/length"
+#define CURRENT_TRIGGER_DEVICE_PATH "/sys/bus/iio/devices/iio:device%d/trigger/current_trigger"
+
+#define SYSFS_TRIGGER_NOW_PATH "/sys/bus/iio/devices/%s/trigger_now"
+#define SYSFS_TRIGGER_NOW_BY_ID_PATH "/sys/bus/iio/devices/trigger%d/trigger_now"
+#define SYSFS_TRIGGER_NAME_PATH "/sys/bus/iio/devices/%s/name"
+#define SYSFS_TRIGGER_SYSFS_ADD_TRIGGER "/sys/bus/iio/devices/iio_sysfs_trigger/add_trigger"
+#define SYSFS_TRIGGER_SYSFS_PATH "/sys/bus/iio/devices/iio_sysfs_trigger"
+
+#define CHANNEL_RAW_PATH "/sys/bus/iio/devices/iio:device%d/%s_raw"
+#define CHANNEL_OFFSET_PATH "/sys/bus/iio/devices/iio:device%d/%s_offset"
+#define CHANNEL_SCALE_PATH "/sys/bus/iio/devices/iio:device%d/%s_scale"
+
+#define CHANNEL_SCAN_ENABLE_PATH "/sys/bus/iio/devices/iio:device%d/scan_elements/%s_en"
+#define CHANNEL_SCAN_INDEX_PATH "/sys/bus/iio/devices/iio:device%d/scan_elements/%s_index"
+#define CHANNEL_SCAN_TYPE_PATH "/sys/bus/iio/devices/iio:device%d/scan_elements/%s_type"
+
+#define SAMPLING_FREQUENCY_DEVICE_PATH "/sys/bus/iio/devices/iio:device%d/sampling_frequency"
+#define SAMPLING_FREQUENCY_BUFFER_PATH "/sys/bus/iio/devices/iio:device%d/buffer/sampling_frequency"
+#define SAMPLING_FREQUENCY_TRIGGER_PATH "/sys/bus/iio/devices/trigger%d/sampling_frequency"
+
+static bool
+craft_filename_path(char *path, size_t size, const char *base, ...)
+{
+    va_list args;
+    int len;
+
+    va_start(args, base);
+    len = vsnprintf(path, size, base, args);
+    va_end(args);
+
+    return len >= 0 && len <= PATH_MAX;
+}
+
+static bool
+write_file(const char *path, const char *format, ...)
+{
+    va_list args;
+    char *content;
+    int fd;
+    size_t len;
+    bool result = true;
+
+    va_start(args, format);
+    vasprintf(&content, format, args);
+    va_end(args);
+
+    fd = open(path, O_WRONLY | O_CLOEXEC);
+    if (fd < 0)
+        result = false;
+
+    if (result) {
+        len = strlen(content);
+        result = write(fd, content, len) == (ssize_t)len;
+    }
+
+    free(content);
+    close(fd);
+
+    return result;
+}
+
+static bool
+check_trigger_now_file(const char *trigger_dir)
+{
+    int len;
+    char path[PATH_MAX + 1];
+    struct stat st;
+
+    len = snprintf(path, sizeof(path), SYSFS_TRIGGER_NOW_PATH, trigger_dir);
+
+    // TODO check permissions of file trigger_now too, not only existence
+    return len > 0 && len <= PATH_MAX && !stat(path, &st);
+}
+
+static bool
+check_trigger_name(const char *trigger_dir, const char *trigger_name)
+{
+    char path[PATH_MAX + 1];
+    char *name;
+    int len;
+    bool result;
+
+    len = snprintf(path, sizeof(path), SYSFS_TRIGGER_NAME_PATH, trigger_dir);
+    if (len < 0 || len > PATH_MAX) {
+        SOL_WRN("Could not open IIO trigger (%s) name on sysfs", trigger_dir);
+        return false;
+    }
+
+    len = sol_util_read_file(path, "%ms", &name);
```

Comment:
I've tried using it, but it is a pain, really. I have to account for empty strings and '\n' on files. =/